man/man1/delta.1
.\"	@(#)delta.1	6.3 (Berkeley) 10/21/04
.tr ~
.tr $%
.if t .tr #\fB\(em\fP
.if n .tr #-
.if n .ds D " -- 
.if t .ds D _
.tr @|
.nr f 0
.bd S B 3
.de SP
.if n .ul
\%[\fB\(em\\$1\fR\\c
.if n .ul 0
\\$2\\$3
..
.de SF
.if n .ul
\%[\fB\(em\\$1\fR]
.if n .ul 0
..
.de ZZ
.hc ^
\fB^...\fR
.hc
..
.de AR
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\\fR
\\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de C1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\\$1
\\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de A1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\fR[\fI\\$2\fR]
\\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de A2
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\fI\\$2\fR
\\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de FI
.PP
.TP 30
\\$1
\\$2
.i0
..
.ds F)  \fB\s-1FILES\s0\fR
.ds W)  \fB\s-1WARNINGS\s0\fR
.ds X)  \fB\s-1EXAMPLES\s0\fR
.ds T)  \fB\s-1TABLE\s0\fR
.ds K)  \fB\s-1DATA KEYWORDS\s0\fR
.ds D)  \fB\s-1DDDDD\s0\fR
.ds M)  \fB\s-1MR\s0\fR
.ds R)  \fB\s-1RELEASE NUMBER\s0\fR
.ds S)  \fB\s-1SCCS\s0\fR
.ds I)  \fB\s-1SID\s0\fR
.de F1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 13
.TP 7
\fB\\$1\fI\\$2\fR
\\$3
.nr f 1
..
.de F2
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 13
.TP 7
\fB\\$1[\fI\\$2\fR]
\\$3
.nr f 1
..
.if n .ds )Q '
.if n .ds )G `
.if t .ds )Q \\(aa
.if t .ds )G \\(ga
.if t .ds )S \\|
.TH DELTA 1 "October 21, 2004"
.SH NAME
delta \- make a delta (change) to an \s-1SCCS\s0 file
.SH SYNOPSIS
.B delta
.SP r SID ]
.SF s
.SF n
.SP g list]
.SP m \%[mrlist] ]
.SP y \%[comment] ]
.SF p
file
.ZZ
.ad
.SH DESCRIPTION
.I Delta
is used to permanently introduce into
the named \*(S) file changes that were made to the file retrieved by
.IR get\^ (1)
(called the
.I g-file,
or generated file).
.PP
.I Delta
makes a delta to each named \*(S) file.
If a directory is named,
.I delta
behaves as though each file in the directory were
specified as a named file,
except that non-\*(S) files
(last component of the pathname does not begin with \fBs.\fR)
and unreadable files
are silently ignored.
If a name of \fB\(em\fR is given, the standard input is read
(see \*(W));
each line of the standard input is taken to be the name of an \*(S) file
to be processed.
.PP
.I Delta
may issue prompts on the standard output depending upon
certain keyletters specified and flags
(see
.IR admin\^ (1))
that may be
present in the \*(S) file
(see
.B \(emm
and
.B \(emy
keyletters below).
.PP
Keyletter arguments
apply independently
to each named file.
.A2 r SID Uniquely
identifies which delta is to be made to the \*(S) file.
The use of this keyletter is necessary only if two or more
outstanding
.IR get\^ s
for editing (\fIget\fP
.B \(eme\c
)
on the same \*(S) file were done by the same person
(login name).
The \*(I) value specified with the
.B \(emr
keyletter can be either the \*(I) specified on the
.I get
command line or the \*(I) to be made as reported by the
.I get
command
(see
.IR get\^ ).
A diagnostic results if the specified \*(I) is ambiguous, or,
if necessary and omitted on the command line.
.AR s Suppresses
the issue, on the standard output, of the created delta's \*(I),
as well as the number of lines inserted, deleted and unchanged in the
\*(S) file.
.AR n Specifies retention of the edited
.I g-file
(normally removed at completion of delta processing).
.A2 g list Specifies
a \fIlist\fR (see
.IR get\^ (1)
for the definition of <\fIlist\fR>)
of deltas which are to be
.I ignored
when the file is accessed at the change level (\*(I))
created by this delta.
.A1 m mrlist If
the \*(S) file has the
.B v
flag set
(see
.IR admin\^ (1))
then a Modification Request (\*(M)) number \fImust\fR be
supplied as the reason for creating the new delta.
.C1 ~ If
.B \(emm
is not used and the standard input is a terminal, the prompt
``\*(M)s?'' is issued on the standard output before the standard input
is read; if the standard input is not a terminal, no prompt is issued.
\fIThe\fB ``\fR\s-1MR\s0s\fB?''\fI prompt always precedes the \fB``\fRcomments\fB?''\fI prompt\fR
(see
.B \(emy
keyletter).
.PP
.C1 ~ \*(M)s
in a list are separated by blanks and/or tab characters.
An unescaped new line character terminates the \*(M) list.
.C1 ~ Note
that if the
.B v
flag has a value
(see
.IR admin\^ (1)),
it is taken to be the name of a program (or shell procedure) which will validate
the correctness of the \*(M) numbers.
If a non-zero exit status is returned from \*(M) number validation program,
.I delta
terminates
(it is assumed that the \*(M) numbers were not all valid).
.A1 y comment Arbitrary
text
used to describe the reason for making the delta.
A null string is considered a valid \fIcomment\fR.
.C1 ~ If
.B \(emy
is not specified and the standard input is a terminal, the prompt
\fB``\fRcomments\fB?''\fR is issued on the standard output before the standard
input is read; if the standard input is not a terminal, no
prompt is issued.
An unescaped new line character terminates the comment text.
.AR p Causes
.I delta
to print (on the standard output) the \*(S) file
differences before and after the delta is applied
in a
.IR diff\^ (1)
format
.RE
.SH FILES
All files of the form
.I ?-file
are explained in the
\fISource Code Control System User's Guide\fP.
The naming convention for these files is also described there.
.RE
.TP "\w`/usr/bin/bdiff\ \ \ `u"
g-file
Existed
before the execution of
.I delta;
removed after completion of
.I delta.
.TP
p-file
Existed
before the execution of
.I delta;
may exist after completion of
.I delta.
.TP
q-file
Created during the execution of
.I delta;
removed after completion of
.I delta.
.TP
x-file
Created during the execution of
.I delta;
renamed to \*(S) file after completion of
.I delta.
.TP
z-file
Created during the execution of
.I delta;
removed during the execution of
.I delta.
.TP
d-file
Created during the execution of
.I delta;
removed after completion of
.I delta.
.TP
/usr/bin/bdiff
Program to compute differences
between the ``gotten'' file and the
.I g-file.
.ad
.PP
.SH WARNINGS
No lines beginning with an \s-1SOH ASCII\s0 character (binary 001)
can be placed in the \*(S) file unless it is escaped.
The character has special meaning to \*(S)
(see
.I sccsfile\c
(5)) and will cause an error.
.PP
A
.I get
of many \*(S) files,
followed by a
.I delta
of those files, should be avoided when the
.I get
generates a large amount of data.
Instead,
multiple
.I "get/delta"
sequences should be used.
.PP
If the standard input
(\fB\(em\fR)
is specified on the
.I delta
command line,
the
.B \(emm
(if necessary)
and
.B \(emy
keyletters
.I must
also be present.
Omission of these keyletters causes an error to occur.
.SH "SEE ALSO"
.na
get(1),
admin(1),
prs(1),
sccshelp(1),
sccsfile(5),
bdiff(1)
.br
.ad
.I "Source Code Control System User's Guide"
by L. E. Bonanni and C. A. Salemi.
.SH DIAGNOSTICS
Use
.IR sccshelp\^ (1)
for explanations.
.tr ~~
.tr $$
.tr @@


usr.bin/sccscmds/cmd/delta.c
# include	"../hdr/defines.h"
# include	"../hdr/had.h"
# include	<sys/file.h>

static char Sccsid[] = "@(#)delta.c	4.12	8/28/99";
USXALLOC();

# ifdef LOGDELTA
char	LogFile[] = "/usr/adm/sccs-log";
int	Logfd;
# endif

char	Diffpgm[] = "/usr/bin/bdiff";
FILE	*Diffin;
int	Debug = 0;
struct packet gpkt;
struct sid sid;
int	num_files;
char	had[26];
char	*ilist, *elist, *glist;
char	*Comments, *Mrs;
int	Domrs;
int	verbosity;
int	Did_id;
long	Szqfile;
char	Pfilename[FILESIZE];
FILE	*Xiop;
int	Xcreate;

main(argc,argv)
int argc;
register char *argv[];
{
	register int i;
	register char *p;
	char c;
	int testmore;
	extern delta();
	extern int Fcnt;

	Fflags = FTLEXIT | FTLMSG | FTLCLN;
	for(i=1; i<argc; i++)
		if(argv[i][0] == '-' && (c=argv[i][1])) {
			p = &argv[i][2];
			testmore = 0;
			switch (c) {

			case 'r':
				if (!p[0]) {
					argv[i] = 0;
					continue;
				}
				chksid(sid_ab(p,&sid),&sid);
				break;
			case 'g':
				glist = p;
				break;
			case 'y':
				Comments = p;
				break;
			case 'm':
				Mrs = p;
				break;
			case 'p':
			case 'n':
			case 's':
				testmore++;
				break;
			default:
				fatal("unknown key letter (cm1)");
			}

			if (testmore) {
				testmore = 0;
				if (*p) {
					sprintf(Error, "value after %c arg (cm7)",c);
					fatal(Error);
				}
			}
			if (had[c - 'a']++)
				fatal("key letter twice (cm2)");
			argv[i] = 0;
		}
		else num_files++;

	if (num_files == 0)
		fatal("missing file arg (cm3)");
	if (!HADS)
		verbosity = -1;
# ifdef LOGDELTA
	Logfd = open(LogFile, O_WRONLY | O_CREAT | O_APPEND, 0666);
# endif
	setsig();
	Fflags &= ~FTLEXIT;
	Fflags |= FTLJMP;
	for (i=1; i<argc; i++)
		if (p=argv[i])
			do_file(p,delta);
# ifdef LOGDELTA
	if (Logfd >= 0)
		close(Logfd);
# endif
	exit(Fcnt ? 1 : 0);
}


delta(file)
char *file;
{
	static int first = 1;
	register char *p;
	int n, linenum;
	char type;
	register int ser;
	extern char had_dir, had_standinp;
	extern char *Sflags[];
	char dfilename[FILESIZE];
	char gfilename[FILESIZE];
	char line[512];
	FILE *gin;
	struct stats stats;
	struct pfile *pp;
	int inserted, deleted, orig;
	int newser;
	int status;
	int diffloop;
	int difflim;

	if (setjmp(Fjmp))
		return;
	if (first) {
		first = 0;
		dohist(file);
	}
	sinit(&gpkt,file,1);
	if (lockit(auxf(gpkt.p_file,'z'),2,getpid()))
		fatal("cannot create lock file (cm4)");
	gpkt.p_reopen = 1;
	gpkt.p_stdout = stdout;
	copy(auxf(gpkt.p_file,'g'),gfilename);
	gin = xfopen(gfilename,0);
	pp = rdpfile(&gpkt,&sid);
	gpkt.p_cutoff = pp->pf_date;
	ilist = pp->pf_ilist;
	elist = pp->pf_elist;

	if (dodelt(&gpkt,&stats,0,0) == 0)
		fmterr(&gpkt);
	if ((ser = sidtoser(&pp->pf_gsid,&gpkt)) == 0 ||
		sidtoser(&pp->pf_nsid,&gpkt))
			fatal("invalid sid in p-file (de3)");
	doie(&gpkt,ilist,elist,glist);
	setup(&gpkt,ser);
	finduser(&gpkt);
	doflags(&gpkt);
	bcopy(&pp->pf_nsid,&gpkt.p_reqsid,sizeof(gpkt.p_reqsid));
	permiss(&gpkt);
	flushto(&gpkt,EUSERTXT,1);
	gpkt.p_chkeof = 1;
	copy(auxf(gpkt.p_file,'d'),dfilename);
	gpkt.p_gout = xfcreat(dfilename,0444);
	while(readmod(&gpkt)) {
		chkid(gpkt.p_line);
		fputs(gpkt.p_line,gpkt.p_gout);
	}
	fclose(gpkt.p_gout);
	orig = gpkt.p_glnno;
	gpkt.p_glnno = 0;
	gpkt.p_verbose = verbosity;
	Did_id = 0;
	while (fgets(line,sizeof(line),gin) != NULL && !chkid(line))
		;
	fclose(gin);
	if (gpkt.p_verbose && (num_files > 1 || had_dir || had_standinp))
		fprintf(gpkt.p_stdout,"\n%s:\n",gpkt.p_file);
	if (!Did_id)
		if (Sflags[IDFLAG - 'a'])
			fatal("no id keywords (cm6)");
		else if (gpkt.p_verbose)
			fprintf(stderr,"No id keywords (cm7)\n");

	/*
	The following while loop executes 'bdiff' on g-file and
	d-file. If 'bdiff' fails (usually because segmentation
	limit it is using is too large for 'diff'), it is
	invoked again, with a lower segmentation limit.
	*/
	difflim = 3500;
	diffloop = 0;
	while (1) {
		inserted = deleted = 0;
		gpkt.p_glnno = 0;
		gpkt.p_upd = 1;
		gpkt.p_wrttn = 1;
		getline(&gpkt);
		gpkt.p_wrttn = 1;
		newser = mkdelt(&gpkt,&pp->pf_nsid,&pp->pf_gsid,
						diffloop,orig);
		diffloop = 1;
		flushto(&gpkt,EUSERTXT,0);
		Diffin = dodiff(auxf(gpkt.p_file,'g'),dfilename,difflim);
		while (n = getdiff(&type,&linenum)) {
			if (type == INS) {
				inserted += n;
				insert(&gpkt,linenum,n,newser);
			}
			else {
				deleted += n;
				delete(&gpkt,linenum,n,newser);
			}
		}
		fclose(Diffin);
		if (gpkt.p_iop)
			while (readmod(&gpkt))
				;
		wait(&status);
		if (status) {		/* diff failed */
			/*
			Check top byte (exit code of child).
			*/
			if (((status >> 8) & 0377) == 32) { /* 'execl' failed */
				sprintf(Error, "cannot execute '%s' (de12)", Diffpgm);
				fatal(Error);
			}
			/*
			Re-try.
			*/
			if (difflim -= 500) {	/* reduce segmentation */
				fprintf(stderr,
			"'%s' failed, re-trying, segmentation = %d (de13)\n",
					Diffpgm,difflim);
				fclose(Xiop);	/* set up */
				Xiop = 0;	/* for new x-file */
				Xcreate = 0;
				/*
				Re-open s-file.
				*/
				gpkt.p_iop = xfopen(gpkt.p_file,0);
				setbuf(gpkt.p_iop,gpkt.p_buf);
				/*
				Reset counters.
				*/
				gpkt.p_slnno = 0;
				gpkt.p_ihash = 0;
				gpkt.p_chash = 0;
				gpkt.p_nhash = 0;
				gpkt.p_keep = 0;
			}
			else
				/* tried up to 500 lines, can't go on */
				fatal("diff failed (de4)");
		}
		else {		/* no need to try again, worked */
			break;			/* exit while loop */
		}
	}
	unlink(dfilename);
	stats.s_ins = inserted;
	stats.s_del = deleted;
	stats.s_unc = orig - deleted;
	if (gpkt.p_verbose) {
		fprintf(gpkt.p_stdout,"%u inserted\n",stats.s_ins);
		fprintf(gpkt.p_stdout,"%u deleted\n",stats.s_del);
		fprintf(gpkt.p_stdout,"%u unchanged\n",stats.s_unc);
	}
	flushline(&gpkt,&stats);
	rename(auxf(gpkt.p_file,'x'),gpkt.p_file);
	if (Szqfile)
		rename(auxf(&gpkt.p_file,'q'),Pfilename);
	else {
		xunlink(Pfilename);
		xunlink(auxf(&gpkt.p_file,'q'));
	}
	clean_up(0);
	if (!HADN) {
		setuid(getuid());
		unlink(gfilename);
	}
}


mkdelt(pkt,sp,osp,diffloop,orig_nlines)
struct packet *pkt;
struct sid *sp, *osp;
int diffloop;
int orig_nlines;
{
	extern long Timenow;
	struct deltab dt;
	char str[128];
	int newser;
	extern char *Sflags[];
	register char *p;
	int ser_inc, opred, nulldel;

	if (!diffloop && pkt->p_verbose) {
		sid_ba(sp,str);
		fprintf(pkt->p_stdout,"%s\n",str);
	}
	sprintf(str,"%c%c00000\n",CTLCHAR,HEAD);
	putline(pkt,str);
	newstats(pkt,str,"0");
	bcopy(sp,&dt.d_sid,sizeof(dt.d_sid));

	/*
	Check if 'null' deltas should be inserted
	(only if 'null' flag is in file and
	releases are being skipped) and set
	'nulldel' indicator appropriately.
	*/
	if (Sflags[NULLFLAG - 'a'] && (sp->s_rel > osp->s_rel + 1) &&
			!sp->s_br && !sp->s_seq &&
			!osp->s_br && !osp->s_seq)
		nulldel = 1;
	else
		nulldel = 0;
	/*
	Calculate how many serial numbers are needed.
	*/
	if (nulldel)
		ser_inc = sp->s_rel - osp->s_rel;
	else
		ser_inc = 1;
	/*
	Find serial number of the new delta.
	*/
	newser = dt.d_serial = maxser(pkt) + ser_inc;
	/*
	Find old predecessor's serial number.
	*/
	opred = sidtoser(osp,pkt);
	if (nulldel)
		dt.d_pred = newser - 1;	/* set predecessor to 'null' delta */
	else
		dt.d_pred = opred;
	dt.d_datetime = Timenow;
	substr(logname(),dt.d_pgmr,0,LNLNAM);
	dt.d_type = 'D';
	del_ba(&dt,str);
	putline(pkt,str);
# ifdef LOGDELTA
	if (Logfd >= 0) {
		char buf[4096], *bp;
		char *cp;

		if (pkt->p_file[0] != '/' && getwd(buf) != NULL)
			strcat(buf, "/");
		else
			buf[0] = '\0';
		strcat(buf, pkt->p_file);
		strcat(buf, ":\n");
		strcat(buf, str + 5);
		bp = buf + strlen(buf);
		for (cp = Comments; *cp;) {
			if (*cp != CTLCHAR)
				*bp++ = *cp++;
			else
				cp += 3;
		}
		*bp++ = '\n';
		write(Logfd, buf, bp - buf);
	}
# endif
	if (ilist)
		mkixg(pkt,INCLUSER,INCLUDE);
	if (elist)
		mkixg(pkt,EXCLUSER,EXCLUDE);
	if (glist)
		mkixg(pkt,IGNRUSER,IGNORE);
	if (Mrs) {
		if (!(p = Sflags[VALFLAG - 'a']))
			fatal("MRs not allowed (de8)");
		if (*p && !diffloop && valmrs(pkt,p))
			fatal("invalid MRs (de9)");
		putmrs(pkt);
	}
	else if (Sflags[VALFLAG - 'a'])
		fatal("MRs required (de10)");
	sprintf(str,"%c%c ",CTLCHAR,COMMENTS);
	putline(pkt,str);
	putline(pkt,Comments);
	putline(pkt,"\n");
	sprintf(str,CTLSTR,CTLCHAR,EDELTAB);
	putline(pkt,str);
	if (nulldel)			/* insert 'null' deltas */
		while (--ser_inc) {
			sprintf(str,"%c%c %s/%s/%05u\n", CTLCHAR, STATS, "00000", "00000", orig_nlines);
			putline(pkt,str);
			dt.d_sid.s_rel -= 1;
			dt.d_serial -= 1;
			if (ser_inc != 1)
				dt.d_pred -= 1;
			else
				dt.d_pred = opred;	/* point to old pred */
			del_ba(&dt,str);
			putline(pkt,str);
			sprintf(str,"%c%c ",CTLCHAR,COMMENTS);
			putline(pkt,str);
			putline(pkt,"AUTO NULL DELTA\n");
			sprintf(str,CTLSTR,CTLCHAR,EDELTAB);
			putline(pkt,str);
		}
	return(newser);
}


mkixg(pkt,reason,ch)
struct packet *pkt;
int reason;
char ch;
{
	int n;
	char str[512];

	sprintf(str,"%c%c",CTLCHAR,ch);
	putline(pkt,str);
	for (n = maxser(pkt); n; n--) {
		if (pkt->p_apply[n].a_reason == reason) {
			sprintf(str," %u",n);
			putline(pkt,str);
		}
	}
	putline(pkt,"\n");
}


putmrs(pkt)
struct packet *pkt;
{
	register char **argv;
	char str[64];
	extern char *Varg[];

	for (argv = &Varg[VSTART]; *argv; argv++) {
		sprintf(str,"%c%c %s\n",CTLCHAR,MRNUM,*argv);
		putline(pkt,str);
	}
}


rdpfile(pkt,sp)
register struct packet *pkt;
struct sid *sp;
{
	char *user;
	struct pfile pf;
	static struct pfile goodpf;
	char line[512];
	int cnt, root;
	FILE *in, *out;

	cnt = -1;
	user = logname();
	bzero(&goodpf,sizeof(goodpf));
	in = xfopen(auxf(pkt->p_file,'p'),0);
	out = xfcreat(auxf(pkt->p_file,'q'),0644);
	root = getuid() == 0;
	while (fgets(line,sizeof(line),in) != NULL) {
		pf_ab(line,&pf,1);
		if (root || equal(pf.pf_user,user)) {
			if (sp->s_rel == 0) {
				if (++cnt) {
					fclose(out);
					fclose(in);
					fatal("missing -r argument (de1)");
				}
				bcopy(&pf,&goodpf,sizeof(pf));
				continue;
			}
			else if (sp->s_rel == pf.pf_gsid.s_rel &&
				sp->s_lev == pf.pf_gsid.s_lev &&
				sp->s_br == pf.pf_gsid.s_br &&
				sp->s_seq == pf.pf_gsid.s_seq) {
					bcopy(&pf,&goodpf,sizeof(pf));
					continue;
			}
		}
		fputs(line,out);
	}
	fflush(out);
	fstat(fileno(out),&Statbuf);
	Szqfile = Statbuf.st_size;
	copy(auxf(pkt->p_file,'p'),Pfilename);
	fclose(out);
	fclose(in);
	if (!goodpf.pf_user[0])
		fatal("not in p-file (de2)");
	return(&goodpf);
}


dodiff(newf,oldf,difflim)
char *newf, *oldf;
int difflim;
{
	register int i;
	int pfd[2];
	FILE *iop;
	extern char Diffpgm[];
	char num[10];

	xpipe(pfd);
	if ((i = fork()) < 0) {
		close(pfd[0]);
		close(pfd[1]);
		fatal("cannot fork, try again (de11)");
	}
	else if (i == 0) {
		close(pfd[0]);
		dup2(pfd[1], 1);
		if (pfd[1] != 1)
			close(pfd[1]);
		for (i = getdtablesize(); i > 4; i--)
			close(i);
		sprintf(num,"%d",difflim);
		execl(Diffpgm,Diffpgm,oldf,newf,num,"-s",0);
		close(1);
		exit(32);	/* tell parent that 'execl' failed */
	}
	else {
		close(pfd[1]);
		iop = fdopen(pfd[0],"r");
		return(iop);
	}
}


getdiff(type,plinenum)
register char *type;
register int *plinenum;
{
	char line[512];
	register char *p;
	int num_lines;
	static int chg_num, chg_ln;
	int lowline, highline;

	if ((p = rddiff(line,sizeof (line))) == NULL)
		return(0);

	if (*p == '-') {
		*type = INS;
		*plinenum = chg_ln;
		num_lines = chg_num;
	}
	else {
		p = linerange(p,&lowline,&highline);
		*plinenum = lowline;

		switch(*p++) {
		case 'd':
			num_lines = highline - lowline + 1;
			*type = DEL;
			skiplines(line,num_lines);
			break;

		case 'a':
			linerange(p,&lowline,&highline);
			num_lines = highline - lowline + 1;
			*type = INS;
			break;

		case 'c':
			chg_ln = lowline;
			num_lines = highline - lowline + 1;
			linerange(p,&lowline,&highline);
			chg_num = highline - lowline + 1;
			*type = DEL;
			skiplines(line,num_lines);
			break;
		}
	}

	return(num_lines);
}


insert(pkt,linenum,n,ser)
register struct packet *pkt;
register int linenum;
register int n;
int ser;
{
	char str[512];

	after(pkt,linenum);
	sprintf(str,"%c%c %u\n",CTLCHAR,INS,ser);
	putline(pkt,str);
	for (++n; --n; ) {
		rddiff(str,sizeof(str));
		putline(pkt,&str[2]);
	}
	sprintf(str,"%c%c %u\n",CTLCHAR,END,ser);
	putline(pkt,str);
}


delete(pkt,linenum,n,ser)
register struct packet *pkt;
register int linenum;
int n;
register int ser;
{
	char str[512];

	before(pkt,linenum);
	sprintf(str,"%c%c %u\n",CTLCHAR,DEL,ser);
	putline(pkt,str);
	after(pkt,linenum + n - 1);
	sprintf(str,"%c%c %u\n",CTLCHAR,END,ser);
	putline(pkt,str);
}


after(pkt,n)
register struct packet *pkt;
register int n;
{
	before(pkt,n);
	if (pkt->p_glnno == n)
		putline(pkt,0);
}


before(pkt,n)
register struct packet *pkt;
register int n;
{
	while (pkt->p_glnno < n) {
		if (!readmod(pkt))
			break;
	}
}


linerange(cp,low,high)
register char *cp;
register int *low, *high;
{
	cp = satoi(cp,low);
	if (*cp == ',')
		cp = satoi(++cp,high);
	else
		*high = *low;

	return(cp);
}


skiplines(lp,num)
register char *lp;
register int num;
{
	for (++num;--num;)
		rddiff(lp,512);
}


rddiff(s,n)
register char *s;
register int n;
{
	register int r;

	if ((r = fgets(s,n,Diffin)) != NULL && HADP)
		fputs(s,gpkt.p_stdout);
	return(r);
}


enter(pkt,ch,n,sidp)
struct packet *pkt;
char ch;
int n;
struct sid *sidp;
{
	char str[32];
	register struct apply *ap;

	sid_ba(sidp,str);
	ap = &pkt->p_apply[n];
	if (pkt->p_cutoff > pkt->p_idel[n].i_datetime)
		switch(ap->a_code) {
	
		case EMPTY:
			switch (ch) {
			case INCLUDE:
				condset(ap,APPLY,INCLUSER);
				break;
			case EXCLUDE:
				condset(ap,NOAPPLY,EXCLUSER);
				break;
			case IGNORE:
				condset(ap,EMPTY,IGNRUSER);
				break;
			}
			break;
		case APPLY:
			fatal("internal error in delta/enter() (de5)");
			break;
		case NOAPPLY:
			fatal("internal error in delta/enter() (de6)");
			break;
		default:
			fatal("internal error in delta/enter() (de7)");
			break;
		}
}


escdodelt()	/* dummy routine for dodelt() */
{
}


clean_up(n)
{
	if (gpkt.p_file[0])
		unlockit(auxf(gpkt.p_file,'z'),getpid());
	xrm(&gpkt);
	xfreeall();
}
